src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.PerAnalyzerState.cs

```diff
@@ -119,58 +121,58 @@ public bool HasPendingSymbolAnalysis(ISymbol symbol)
+            public async Task<AnalyzerStateData> TryStartAnalyzingSymbolAsync(ISymbol symbol, CancellationToken cancellationToken)
             {
-                return TryStartProcessingEntity(symbol, _pendingSymbols, _analyzerStateDataPool, out state);
+                return await TryStartProcessingEntityAsync(symbol, _pendingSymbols, _analyzerStateDataPool, cancellationToken).ConfigureAwait(false);
             }
 
-            public void MarkSymbolComplete(ISymbol symbol)
+            public async Task MarkSymbolCompleteAsync(ISymbol symbol, CancellationToken cancellationToken)
             {
-                MarkEntityProcessed(symbol, _pendingSymbols, _analyzerStateDataPool);
+                await MarkEntityProcessedAsync(symbol, _pendingSymbols, _analyzerStateDataPool, cancellationToken).ConfigureAwait(false);
             }
 
-            public bool TryStartAnalyzingDeclaration(SyntaxReference decl, out DeclarationAnalyzerStateData state)
+            public async Task<DeclarationAnalyzerStateData> TryStartAnalyzingDeclarationAsync(SyntaxReference decl, CancellationToken cancellationToken)
             {
-                return TryStartProcessingEntity(decl.GetSyntax(), _pendingDeclarations, _declarationAnalyzerStateDataPool, out state);
+                return await TryStartProcessingEntityAsync(decl.GetSyntax(), _pendingDeclarations, _declarationAnalyzerStateDataPool, cancellationToken).ConfigureAwait(false);
             }
 
-            public bool IsDeclarationComplete(SyntaxNode decl)
+            public async Task<bool> IsDeclarationCompleteAsync(SyntaxNode decl, CancellationToken cancellationToken)
             {
-                return IsEntityFullyProcessed(decl, _pendingDeclarations);
+                return await IsEntityFullyProcessedAsync(decl, _pendingDeclarations, cancellationToken).ConfigureAwait(false);
             }
 
-            public void MarkDeclarationComplete(SyntaxReference decl)
+            public async Task MarkDeclarationCompleteAsync(SyntaxReference decl, CancellationToken cancellationToken)
             {
-                MarkEntityProcessed(decl.GetSyntax(), _pendingDeclarations, _declarationAnalyzerStateDataPool);
+                await MarkEntityProcessedAsync(decl.GetSyntax(), _pendingDeclarations, _declarationAnalyzerStateDataPool, cancellationToken).ConfigureAwait(false);
             }
 
-            public bool TryStartSyntaxAnalysis(SyntaxTree tree, out AnalyzerStateData state)
+            public async Task<AnalyzerStateData> TryStartSyntaxAnalysisAsync(SyntaxTree tree, CancellationToken cancellationToken)
             {
                 Debug.Assert(_lazyPendingSyntaxAnalysisTrees != null);
-                return TryStartProcessingEntity(tree, _lazyPendingSyntaxAnalysisTrees, _analyzerStateDataPool, out state);
+                return await TryStartProcessingEntityAsync(tree, _lazyPendingSyntaxAnalysisTrees, _analyzerStateDataPool, cancellationToken).ConfigureAwait(false);
```

Comment:
`lazyPendingSyntaxAnalysisTree` is only initialized for the Compilation started event, and should happen prior to any analysis, so it still holds true here.